sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecutionSuite.scala

```diff
@@ -68,4 +84,74 @@ class MicroBatchExecutionSuite extends StreamTest with BeforeAndAfter {
       CheckNewAnswer((25, 1), (30, 1))   // This should not throw the error reported in SPARK-24156
     )
   }
+
+  test("SPARK-30143: stop waits until timeout if blocked") {
+    val latch = new CountDownLatch(1)
+    BlockOnStopSourceProvider.setLatch(latch)
+    val sq = spark.readStream.format(BlockOnStopSourceProvider.getClass.getName)
+      .load()
+      .writeStream
+      .format("console")
+      .start()
+    failAfter(60.seconds) {
+      val startTime = System.nanoTime()
+      withSQLConf(SQLConf.STREAMING_STOP_TIMEOUT.key -> "2000") {
+        intercept[TimeoutException] {
+          sq.stop()
+        }
+      }
+      val duration = (System.nanoTime() - startTime) / 1e6
+      assert(duration >= 2000 * 1e6,
+        s"Should have waiter more than 2000 millis, but waited $duration millis")
+
+      latch.countDown()
+      withSQLConf(SQLConf.STREAMING_STOP_TIMEOUT.key -> "0") {
+        sq.stop()
+      }
+    }
+  }
+}
+
+object BlockOnStopSourceProvider extends StreamSourceProvider with TableProvider {
```

Comment:
Would be good to document this (and maybe the other ones) to say that these are for testing stream termination delays by having them not stop until the latch is activated.